zephyr/patches/zephyr/k-busy-wait-ns.patch

```diff
@@ -0,0 +1,16 @@
+diff --git a/include/zephyr/kernel.h b/include/zephyr/kernel.h
+index 614b2eab1b9..e9752bb0fdb 100644
+--- a/include/zephyr/kernel.h
++++ b/include/zephyr/kernel.h
+@@ -617,6 +617,11 @@ __syscall int32_t k_usleep(int32_t us);
+  */
+ __syscall void k_busy_wait(uint32_t usec_to_wait);
+
++static inline void k_busy_wait_ns(uint32_t nsec_to_wait)
++{
++       k_busy_wait(k_cyc_to_us_ceil32(k_ns_to_cyc_floor32(nsec_to_wait)));
```

Comment:
I'm concerned about the amount of rounding we do here. For us (if my math is right), the evaluation would be something like this:

```
k_busy_wait_ns(300)
= k_busy_wait(k_cyc_to_us_ceil32(k_ns_to_cyc_floor32(300)))
= k_busy_wait(k_cyc_to_us_ceil32(240))
= k_busy_wait(k_cyc_to_us_ceil32(240))
= k_busy_wait(1)
```

I think if we want a delay function at this accuracy, we need to implement `k_busy_wait_ns` similar to how `k_busy_wait` is implemented, where we get the cycle count and then delay that many NOPs. We could even move `k_busy_wait` and `k_busy_wait_ns` to use the same underlying function (that delays for a certain cycle or loop count)

https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/busy_wait.c#L12

Reply:
Ok, makes sense. I would really prefer we just use `k_busy_wait(1)` then. IMO it doesn't make sense to create a patch we can't upstream just to use a function we might implement later. Let's either:
- do the full implementation now and upstream
- use `k_busy_wait(1)` and add a comment that the true delay required is 300ns. We can move back to a NS delay function once one exists

Reply:
@danieldegrasse - the reason I suggested to make a workaround here was because I would be doing the upstream work to add support. I think it's fine.

Reply:
Ah ok, if we are going to upstream it that's fine with me then, just wasn't clear from your initial comment.